### DISCUSSION TITLE
SSL_read fail with err_code SSL_ERROR_WANT_READ

### DISCUSSION THREAD

**Discussion by sasakiakaya** [openssl/openssl#23608] [Q&A] [answered]
I use SSL in COAP application (COAP -> DTLS 1.2 -> UDP). SSL_read has high failure frequency with SSL_ERROR_WANT_READ after handshake, and UDP socket actually received SSL application data. This failure caused COAP connect fail.
I try to ignore SSL_ERROR_WANT_READ, and COAP connect success. But I'm concerned that this issue will affect performance. So I wonder what cause this issue and how to solve it.

kernel version: 5.15.119
openssl version: 1.1.1t

UDP payload from socket when SSL_read fail:
0x17 0xfe 0xfd 0x0 0x1 0x0 0x0 0x0 0x0 0x0 
0x1 0x0 0x18 0x0 0x1 0x0 0x0 0x0 0x0 0x0 
0x1 0x9b 0x87 0x8a 0xee 0x36 0x42 0xf0 0x68 0x8f 
0x96 0x87 0xac 0x95 0xcb 0xe9 0x72

**Comment by kroeckx** [comment, marked as answer]
SSL_ERROR_WANT_READ means that at this time no data is available and you should wait for more data. Please the SSL_get_error() manpage. You can wait for more data with something like the select() function.

**Reply by sasakiakaya** [reply]
But I actually received data from socket, what's the mean by "no data"? Is that mean SSL reveive buffer has not update the data from socket?

**Reply by kroeckx** [reply]
A TLS packet can be multiple TCP packets, and a TLS packet can only be processed if it's completely received. So if you've received part of a TLS packet, we'll return that more data is needed with the SSL_ERROR_WANT_READ code.

An other reason that you might receive data on the socket but you get no data is that TLS might send data that is not application data, for instance a session resumption ticket, or a key update.

**Reply by sasakiakaya** [reply]
OK, understood. Thanks for your answer.
